// grunt-cms-site-scaffold/Gruntfile.js
/*
 * grunt-cms-site-scaffold
 * https://github.com/Banno/grunt-cms-site-scaffold
 *
 * Copyright (c) 2013 Jack Viers
 * Licensed under the BANNO license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    email: "noreply@banno.com",
    password: "password",
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>',
      ],
      options: {
        jshintrc: '.jshintrc',
      },
    },
    server: {

    },
    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp'],
    },

    // Configuration to be run (and then tested).
    cms_site_scaffold: {
      default_options: {
        options: {
        },
        src: ["test/fixtures/artemis.local.json"]
      },
      custom_options: {
        options: {
          ignoreMissing: true
        },
        files: [
          {src: ["test/fixtures/artemis.local.json", "test/fixtrues/nonexistant.json"]}
        ]
      },
    },
    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js'],
    },

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');
  grunt.loadNpmTasks('grunt-bump');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'server', 'cms_site_scaffold', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);

};
