// grunt-cms-upload/Gruntfile.js
/*
 * grunt-cms-upload
 * https://github.com/Banno/grunt-cms-upload
 *
 * Copyright (c) 2013 Banno LLC. All rights reserved.
 * Licensed under the BANNO license.
 */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    email: 'foo@banno.com',
    password: 'barbarosa',
    jshint: {
      all: [
        'Gruntfile.js',
        'tasks/*.js',
        '<%= nodeunit.tests %>',
      ],
      options: {
        jshintrc: '.jshintrc',
      },
    },
    server: {

    },
    // Before generating any new files, remove any previously-created files.
    clean: {
      tests: ['tmp'],
    },
    
    // Configuration to be run (and then tested).
    cms_upload: {
      default_options: {
        options: {
          ignoreMissing: false
        },
        files: [
          {src:['test/fixtures/image-001.jpg'], dest:''}
        ]
        //src:['test/fixtures/image-001.jpg']
      },
      custom_options: {
        options: {
          ignoreMissing: true
        },
        files: [
          {src:['test/fixtures/image-001.jpg'], dest:''},
          {src:['test/fixtures/apollo.zip'], dest:''}
        ]
        
      },
    },

    // Unit tests.
    nodeunit: {
      tests: ['test/*_test.js'],
    },

  });

  // Actually load this plugin's task(s).
  grunt.loadTasks('tasks');

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-clean');
  grunt.loadNpmTasks('grunt-contrib-nodeunit');
  grunt.loadNpmTasks('grunt-bump');

  // Whenever the "test" task is run, first clean the "tmp" dir, then run this
  // plugin's task(s), then test the result.
  grunt.registerTask('test', ['clean', 'server', 'cms_upload', 'nodeunit']);

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'test']);

};
